broker/src/serve_tasks.rs

```diff
@@ -99,6 +149,73 @@ fn wait_get_statuscode<S>(vec: &Vec<S>, block: &HowLongToBlock) -> StatusCode {
     }
 }
 
+async fn wait_for_results_for_task_stream<'a, M: Msg, I: PartialEq>(results: &'a mut HashMap<AppOrProxyId, M>, block: &'a HowLongToBlock, mut new_result_rx: Receiver<M>, filter: &'a MsgFilterNoTask<'a>, mut deleted_task_rx: Receiver<MsgId>, task_id: &'a MsgId)
+-> impl Stream<Item = Result<Event,Infallible>> + 'a
+where M: Clone + HasWaitId<I>
+{
+    let wait_until =
+        time::Instant::now() + block.wait_time.unwrap_or(time::Duration::from_secs(31536000));
+    trace!(
+        "Now is {:?}. Will wait until {:?}",
+        time::Instant::now(),
+        wait_until
+    );
+    let mut running = true;
+    let stream = async_stream::stream! {
+    while usize::from(block.wait_count.unwrap_or(0)) > results.len()
+        && time::Instant::now() < wait_until
+        && running {
+        trace!(
+            "Items in vec: {}, time remaining: {:?}",
+            results.len(),
+            wait_until - time::Instant::now()
+        );
+            tokio::select! {
+                _ = tokio::time::sleep_until(wait_until) => {
+                    debug!("SSE: Wait expired.");
+                    yield Ok(Event::default()
+                        .event(SseEventType::WaitExpired)
+                        .data("{}"));
+                    running = false;
+                },
+                result = new_result_rx.recv() => {
+                    match result {
+                        Ok(req) => {
+                            if filter.matches(&req) {
+                                let previous = results.insert(req.get_from().clone(), req.clone());
+                                let event_type = match previous {
+                                    Some(_) => SseEventType::UpdatedResult,
+                                    None => SseEventType::NewResult
+                                };
+                                yield Ok(Event::default()
+                                    .event(event_type)
+                                    .json_data(&req)
+                                    .unwrap());
+                            }
+                        },
+                        Err(e) => { panic!("Unable to receive from queue new_result_rx: {}", e); }
```

Comment:
Ouch, is here a panic necessary?

Reply:
Yes -- this would be a serious internal error. If we want to change this, we should look around the whole application and think of some kind of error handling (although I'm not sure how Beam would recover) in another PR.